### DISCUSSION TITLE
Redis failovers when scheduling jobs

### DISCUSSION THREAD

**Discussion by franzliedke** [sidekiq/sidekiq#4990] [Q&A]
Sidekiq has some special handling for errors related to Redis cluster fail-overs, where connections to previous masters are now connected to a read-only replica. (Thanks for merging my related PR!)

As far as I'm aware, this special handling all happens in the [`Sidekiq.redis` method](https://github.com/mperham/sidekiq/blob/ab90830e85e02304ac39dc29f8684c75564726de/lib/sidekiq.rb#L94-L113).

We are still seeing those errors, however, when scheduling jobs.

<details>
<summary>Stacktrace</summary>

```
Redis::CommandError: READONLY You can't write against a read only replica.
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis/client.rb:216:in `call_pipelined'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis/client.rb:170:in `block in call_pipeline'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis/client.rb:313:in `with_reconnect'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis/client.rb:168:in `call_pipeline'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:2490:in `block in multi'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:69:in `block in synchronize'
  from monitor.rb:202:in `synchronize'
  from monitor.rb:202:in `mon_synchronize'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:69:in `synchronize'
  from vendor/ruby/2.7.0/gems/redis-4.2.5/lib/redis.rb:2483:in `multi'
  from vendor/ruby/2.7.0/gems/sidekiq-6.2.1/lib/sidekiq/client.rb:189:in `block in raw_push'
  from vendor/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:63:in `block (2 levels) in with'
  from vendor/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:62:in `handle_interrupt'
  from vendor/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:62:in `block in with'
  from vendor/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:59:in `handle_interrupt'
  from vendor/ruby/2.7.0/gems/connection_pool-2.2.5/lib/connection_pool.rb:59:in `with'
  from vendor/ruby/2.7.0/gems/sidekiq-6.2.1/lib/sidekiq/client.rb:188:in `raw_push'
  from vendor/ruby/2.7.0/gems/sidekiq-6.2.1/lib/sidekiq/client.rb:74:in `push'
  from vendor/ruby/2.7.0/gems/sidekiq-6.2.1/lib/sidekiq/worker.rb:240:in `client_push'
  from vendor/ruby/2.7.0/gems/sidekiq-6.2.1/lib/sidekiq/worker.rb:200:in `perform_async'
  from app/models/my_model.rb:49:in `schedule_sync!'
```

</details>

Unless I'm missing something in this stacktrace, the scheduling circumvents `Sidekiq.redis` in favor of the `Client`'s [`#raw_push` method](https://github.com/mperham/sidekiq/blob/ab90830e85e02304ac39dc29f8684c75564726de/lib/sidekiq/client.rb#L187-L194), which directly access the Redis connection pool.

What is the purpose of this shortcut? Would you accept a PR that extracts the reconnect logic to a standalone helper that could be re-used in both/more places?

**Comment by mperham** [comment]
Yes, it needs that same logic. PR welcome.

**Reply by franzliedke** [reply]
I will happily send that PR. Would you like `raw_push` to use `Sidekiq.redis` or is it "raw" on purpose?

**Comment by mperham** [comment]
It’s on purpose.

**Reply by franzliedke** [reply]
Care to explain?

**Comment by mperham** [comment]
The client has to push dynamically to different pools which target
different Redises in order to support sharding.

**Comment by franzliedke** [comment]
Status update: We're currently experimenting with a monkey-patch in our codebase, but aim to contribute it upstream once we are sure it works and solves the problem for us. 😃

**Reply by franzliedke** [reply]
We're having small troubles because `sidekiq/testing` also monkey-patches `#raw_push` 😅
